aQute.libg/src/aQute/libg/uri/URIUtil.java

```diff
@@ -41,4 +44,84 @@ public static URI resolve(URI baseURI, String reference) throws URISyntaxExcepti
 		return baseURI.resolve(refURI);
 	}
 
+	private static final URI EMPTYURI = URI.create("");
+
+	/**
+	 * Attempts to fetch a path on the file system for the given URI string.
+	 * Tries a few more tricks than the standard method of
+	 * <tt>Paths.get(new URI(uriString))</tt> - it can handle plain paths, and
+	 * also nested URI pseudo-schemes like <tt>reference:</tt> and
+	 * <tt>jar:</tt>. Examples:
+	 * <ul>
+	 * <li>/path/to/file => /path/to/file</li>
+	 * <li>reference:file:/path/to/file => /path/to/file</li>
+	 * <li>jar:file:/path/to/file.jar!/some/contained/element =>
+	 * /path/to/file.jar</li>
+	 * <li>http://server/path/to/file => <tt>null</tt></li>
+	 * </ul>
+	 *
+	 * @param uriString The string containing the URI for which we are
+	 *            attempting to construct a path.
+	 * @return The method's best guess as to which file on the local filesystem
+	 *         this URI refers, or <tt>null</tt> if it's invalid or not a local
+	 *         filesystem URI.
+	 */
+	public static Path fromURI(String uriString) {
+		if (uriString == null || uriString.isEmpty()) {
+			return null;
+		}
+
+		try {
+			URI uri = resolve(EMPTYURI, uriString);
+			while (uriString != null && !uriString.isEmpty()) {
+
+				final String scheme = uri.getScheme();
+				if (scheme == null) {
+					try {
+						return Paths.get(uriString);
+					} catch (InvalidPathException e) {
+						e.printStackTrace();
```

Comment:
This should not be in production code. If it is an unrecoverable error that we wish to communicate, then we need to throw a meaningful exception. Printing to System.out can be lost, in Bndtools for example.

Reply:
Sorry, a relic from my debugging. It should be ignored.

Reply:
`printStackTrace()` is gone now.